### PR TITLE
[Diffusion] Add rank-local mmap storage for DLO

### PR DESCRIPTION
**Pull request by hsliuustc0106**
## Summary

Refs #6195.

This PR makes node-local mmap a DLO host-weight storage option independent of the AllGather transfer protocol.

- retain compatible checkpoint tensors as immutable file-backed sources for rank-local DLO
- stream complete blocks through two bounded pinned host staging slots per worker
- preserve the existing DP-sharded AllGather path and standard-loader fallbacks
- add MiniMax-H3 TP1 mmap support with grouped-QKV layout adaptation
- document the storage/transfer/orchestration boundaries and add a DP2 lifecycle mode
- make no scheduler, executor, worker-dispatch, or request-routing changes

## Motivation

DLO currently has two transfer modes with different host-memory behavior:

| Mode | Persistent host storage | Per-block transfer |
| --- | --- | --- |
| AllGather | a private `1 / group_size` shard per rank | shard H2D + AllGather |
| no-AllGather, before this PR | a private complete model per rank | complete-block H2D |
| no-AllGather, this PR | node-shared file-backed pages + bounded private staging | complete-block H2D |

no-AllGather requires every rank to transfer a complete block, but it does not require every rank to retain a persistent anonymous copy of the same immutable checkpoint. Processes on one node can fault the same clean safetensors pages through the OS page cache.

This is deliberately a storage/lifecycle change. Cross-replica admission and load balancing remain outside DLO and can be owned by vLLM Router, Dynamo, or another deployment layer.

### Shared selection policy

The diffusion loader and DLO backend call the same `should_select_dlo_mmap()` policy. The loader skips ordinary weight materialization only when the backend will enter the mmap path.

Rank-local mmap initially requires:

- an explicitly compatible model;
- TP=1;
- HSDP disabled; and
- no online quantization.

Other no-AllGather layouts retain the regular loader.

### Exact component source mapping

The mmap loader indexes each declared component source, including safetensors index files, prefixes keys into the pipeline namespace, and applies an optional checkpoint-key remap. This avoids recursively conflating unrelated checkpoint partitions.

### Retained source and bounded staging

Each rank-local hook retains detached checkpoint views and replaces runtime parameters with offload placeholders. All streamed and resident DiT layers in one worker share two host staging slots sized to the largest block.

A prefetch waits for prior slot use, packs one block, applies any declared layout adapter, copies the complete block to the rotating device slot, and records an event protecting host-slot reuse. The safetensors mappings remain open until DLO is disabled.

CPU packing is synchronous in this first implementation. The pinned H2D transfer can overlap with current-block compute.

### MiniMax-H3

MiniMax-H3 explicitly opts into TP1 mmap loading. Its checkpoint interleaves Q/K/V rows by query group, while the fused runtime linear expects contiguous Q, K, then V rows. A per-parameter adapter applies the same reorder as the regular loader while packing one bounded block; the file-backed source remains unchanged.

Quantized MiniMax-H3 continues to use the regular loader.

## Usage

```bash
vllm serve /path/to/model --omni \
  --enable-distributed-layerwise-offload \
  --data-parallel-size 2 \
  --dlo-no-use-allgather
```

For an explicitly compatible TP1 model, this selects rank-local mmap automatically. Unsupported models and parallel layouts retain existing behavior.

## Scope and limitations

- Multi-rank AllGather behavior is unchanged.
- With an effective DLO group size of one, AllGather configuration now correctly uses rank-local mmap staging, closing a pre-existing loader/backend mismatch that could leave parameters on `meta`.
- no-AllGather still performs a complete-block H2D copy per rank.
- The sharing boundary is one OS node, not the cluster.
- The first scope is the streamed DiT and planned DiT submodules; ancillary encoders/VAEs and fallback paths may retain private host weights.
- TP, HSDP, and online quantization keep their normal loader contracts.
- This PR changes no request scheduling or DP orchestration code.

### CPU and static checks

```text
234 passed
ruff check: passed
ruff format --check: passed
git diff --check: passed
```

The suite covers the DLO backend, diffusion loader, MiniMax-H3 contracts and quantization behavior, diffusion-engine regressions, and entrypoint configuration.

### MiniMax-H3 DP2 no-AllGather E2E

Exact final patch, 2x NVIDIA L20X, MiniMax-H3 FL2VA:

```text
data_parallel_size=2
tensor_parallel_size=1
dlo_use_allgather=false
535/535 checkpoint tensors mapped
52 streamed blocks across 2 hook groups
2 x 1.23 GiB pinned host staging slots per worker
initial GPU model memory: 3.25 GiB per worker
frames: [107, 256, 256, 3]
audio: [1, 2, 142400]
peak GPU memory: 13,226 MiB
request wave: 41.57 s
active_children after shutdown: []
```

A page-sharing inspection of the same path found 13 transformer mappings covering approximately 63.2 GiB. During denoising, each process reported roughly 50–51 GiB `Shared_Clean`, approximately half-sized PSS, and zero `Private_Dirty` for those mappings, confirming shared clean checkpoint pages rather than per-process anonymous full-model copies.

### CODE REVIEW

File: vllm_omni/diffusion/offloader/offload_plan.py [hunk 1]
```diff
@@ -60,15 +60,46 @@ def get_offload_plan(pipeline: nn.Module) -> OffloadPlan | None:
...
+    tensor_parallel_size: int,
+    use_hsdp: bool,
+) -> bool:
+    """Whether loader and backend should enter DLO's mmap path.
+
+    AllGather preserves its existing fail-closed topology validation inside
+    the mmap loader. The extra TP/HSDP gate applies to rank-local mmap only.
+    """
+    rank_local_compatible = tensor_parallel_size == 1 and not use_hsdp
+    return mmap_supported and not has_online_quant and (dlo_use_allgather or rank_local_compatible)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/offload_plan.py:110]: With `dlo_use_allgather=True` and `dp_size=1` this now selects mmap — the old gate in `enable()` required `dp_size > 1`, so that config skipped `load_weights` in the loader but never ran `_load_weights_via_mmap()` and was left on meta params. Nice that this closes that hole, but since the PR body says AllGather behavior is unchanged, maybe add a sentence there acknowledging the `dp_size=1` corner.

File: vllm_omni/diffusion/model_loader/diffusers_loader.py [hunk 2]
```diff
@@ -363,22 +363,39 @@ def load_model(
...
+                # The loader and backend share should_select_dlo_mmap() so
+                # skipping load_weights always selects mmap in enable().
+                from vllm_omni.diffusion.offloader.offload_plan import (
+                    should_select_dlo_mmap,
+                    supports_mmap_loading,
+                )
 
                 _dist_offload = getattr(self.od_config, "enable_distributed_layerwise_offload", False)
                 _use_ag = getattr(self.od_config, "dlo_use_allgather", True)
                 _has_online_quant = self._has_online_quant(model)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/model_loader/diffusers_loader.py:381]: `enable()` computes `has_online_quant` by scanning `quant_method.uses_meta_device` over `pipeline.modules()`, while the loader uses `self._has_online_quant(model)`. If the two ever disagree, `load_weights` is skipped but mmap is not selected, leaving the model on meta params. Pre-existing for AllGather — since both paths now hinge on this, could the detection live in one shared helper?

File: vllm_omni/diffusion/offloader/distributed_layerwise_backend.py [hunk 3]
```diff
@@ -952,27 +1200,33 @@ def _load_weights_via_mmap(self, pipeline: nn.Module, modules) -> None:
...
             # Filter out params inside submodules that will be loaded later
             # by _load_module_weights_from_mmap (non-block submodules)
             dit_param_names = set()
-            for dit_module in modules.dits:
-                for pname, _ in dit_module.named_parameters():
-                    dit_param_names.add(pname)
+            for dit_name, dit_module in zip(modules.dit_names, modules.dits):
+                for pname, _ in chain(dit_module.named_parameters(), dit_module.named_buffers()):
+                    dit_param_names.add(f"{dit_name}.{pname}")
             truly_missing = [n for n in remaining_meta if n in dit_param_names]
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/distributed_layerwise_backend.py:1231]: This only warns, so an opt-in model whose checkpoint keys do not match runtime names (no `_remap_ckpt_key` to catch it) fails later and less clearly — via the CPU-device check in `_collect_mmap_sources` under rank-local, or cryptically under AllGather. Any reason not to raise when `truly_missing` is non-empty for explicitly opted-in models?

File: vllm_omni/diffusion/offloader/base.py [hunk 4]
```diff
@@ -32,6 +32,8 @@ class OffloadConfig:
     dlo_use_allgather: bool = True
     dlo_resident_layers: int = 0  # leading DiT layers kept on device
     model_path: str | None = None  # checkpoint path for mmap weight loading
+    # Appended to preserve the positional order of the pre-existing fields.
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/base.py]: why we need to add tp size here ? is seems not related

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 5]
```diff
@@ -532,13 +532,10 @@ class MiniMaxH3Pipeline(
         encoder_block_attrs={"text_encoder": ("vision.blocks", "text_model.layers")},
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:532]: for other models, what do we change to adapt to this feature?

File: vllm_omni/diffusion/offloader/offload_plan.py (hunk 1 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/offloader/offload_plan.py:110]: Thanks — clarified in both the PR body and the design doc. More precisely, the corner is an effective DLO group size of one, not simply `DP=1`: with `DP=1, SP>1`, DLO can still AllGather over the SP group. Multi-rank AllGather behavior is unchanged; an effective group size of one now uses rank-local mmap staging and closes the previous loader/backend mismatch that could leave parameters on `meta`.

File: vllm_omni/diffusion/model_loader/diffusers_loader.py (hunk 2 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/model_loader/diffusers_loader.py:381]: Done in `76e257b0`. I added shared `has_online_quantization()` policy in `offload_plan.py`; both `DiffusersPipelineLoader._has_online_quant()` and `DistributedLayerwiseOffloadBackend.enable()` now delegate to it. A regression test covers both the false and `uses_meta_device=True` cases.

File: vllm_omni/diffusion/offloader/distributed_layerwise_backend.py (hunk 3 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/offloader/distributed_layerwise_backend.py:1231]: Agreed. In `76e257b0`, explicitly opted-in pipelines (`_supports_mmap_loading=True`) now raise a clear `RuntimeError` immediately when expected DiT tensors remain on `meta`, including the count and first missing names. The warning is retained only for legacy remap-based mmap models, and a regression test covers an incomplete explicit component source.

File: vllm_omni/diffusion/offloader/base.py (hunk 4 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/offloader/base.py]: The TP size is needed for host-storage selection, not for creating the DLO AllGather group. Rank-local mmap bypasses TP-aware `weight_loader` callbacks, so no-AllGather with `TP>1` must retain the standard loader. The loader reads `parallel_config` directly, while the backend receives `OffloadConfig`; carrying the configured TP size lets both call the same `should_select_dlo_mmap()` policy. I added this rationale next to the field in `1c3d5979`.

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 5 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py:532]: I added an explicit “Adding mmap support to another model” contract to the DLO design doc in `1c3d5979`. In short: provide exact safetensors `weights_sources` prefixes; opt in with `_supports_mmap_loading=True` when names match or provide `_remap_ckpt_key` when they do not; attach `mmap_weight_transform` for TP1 layout conversions; and test mapping completeness, equivalence with the regular loader, and TP/HSDP/online-quant fallbacks. The opt-in flag alone now fails fast if any expected DiT tensor is unmatched.

File: vllm_omni/diffusion/offloader/base.py (hunk 4 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/offloader/base.py]: why we only care about TP?

**Comment by hsliuustc0106** [comment]
close this PR since we have updated the design in #6195
